returnn/torch/data/pipeline.py

```diff
@@ -309,6 +311,105 @@ def __iter__(self):
+    a dataset yielding lists of these sequences, i.e. batches. Can be used instead of
+    the default ordered batching via the `custom_batching` config key.
+
+    Sequences are grouped into a distinct set of length buckets. Each bucket has a
+    limit of how many sequences it can buffer before being sent off to training.
+
+    Note, that batches are not yet merged into a single (padded) data array here,
+    this happens in 'collate_batch()'.
+    """
+
+    def __init__(
+        self, dataset: torch.utils.data.IterableDataset, *, buckets: Sequence[Tuple[int, int]], length_key: str
+    ):
+        """
+        :param dataset: dataset to apply bucket batching to
+        :param buckets: Bucket configuration as tuples of seq length and max number of seqs in that bucket.
+            Segments longer than the largest size limit configured in the buckets are dropped. To avoid dropping
+            any segments make sure your largest bucket allows segments larger than your longest training segment.
+        :param length_key: data key to take as length measure
+        """
+        self._dataset = dataset
+        self._length_key = length_key
+
+        assert buckets, "empty bucket batching configuration"
+        if not all(size > 0 and max_seqs > 0 for size, max_seqs in buckets):
+            raise ValueError(f"bucket sizes and max seqs in bucket must be positive")
+        self._max_seq_lens, self._max_bucket_sizes = zip(*sorted(buckets, key=lambda b: b[0]))
+        assert len(set(self._max_seq_lens)) == len(self._max_seq_lens), "seq len boundaries must all be unique"
+
+    def __iter__(self):
+        """:return: generator applying bucket ordering on the data"""
+        buckets: List[List[Dict[str, numpy.ndarray]]] = [[] for _ in range(len(self._max_seq_lens))]
+
+        for data_dict in self._dataset:
+            data_dict: Dict[str, numpy.ndarray]
+            data_size = data_dict[self._length_key].shape[0]
+            bucket_idx = bisect.bisect_left(self._max_seq_lens, data_size)
+            if bucket_idx >= len(self._max_seq_lens):
+                # seg is too long, drop it
+                continue
+            buckets[bucket_idx].append(data_dict)
+            if len(buckets[bucket_idx]) >= self._max_bucket_sizes[bucket_idx]:
+                yield buckets[bucket_idx]
+                buckets[bucket_idx] = []
+
+        non_empty_buckets = [b for b in buckets if b]
+        yield from non_empty_buckets
+
+
+def get_batching_iterable_dataset_from_config(
+    *, dataset: torch.utils.data.IterableDataset, config: Config, train: bool
+) -> torch.utils.data.IterableDataset:
+    """
+    Batches the segments in the given dataset given the config settings.
+
+    :param dataset: dataset whose segments to group into batches
+    :param config: RETURNN config
+    :param train: whether we are in training or in inference mode
+    """
+    custom_batching = config.typed_value("custom_batching", None)
```

Comment:
I thought a bit about the config option name. I don't really like the `custom_` prefix. This is not more custom than anything else really. E.g. this is just like `optimizer`. We also don't call this `custom_optimizer`. It's up to the user how "custom" this is.

So, just `"batching"` makes more sense. However, for `"batching"`, we have the problem that this conflicts with the already existing `"batching"` option, used by `Dataset`:
```python
        set_or_remove("seq_ordering", config.value("batching", None))
```
E.g. when the user has `batching = "laplace:..."` in the config.
This is config option is actually quite related.
But I don't think we can reuse this option here. This would be too confusing (just to figure out where this config option is actually used...).

So, how else could this be called? `batching_method`?

Reply:
Ah, also, this here is now PyTorch specific. So maybe call this `torch_batching`?